runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+	value interpreter.Value,
+	refType interpreter.EphemeralReferenceDynamicType,
+) bool {
+
+	referenceValue, ok := value.(*interpreter.EphemeralReferenceValue)
+	if !ok {
+		return false
+	}
+
+	return valueConformsToDynamicType(*referenceValue.ReferencedValue(), refType.InnerType())
+}
+
+// valueConformsToSemaType checks whether a value conforms to a given semantic type.
+// Returns `true` if the value and the type matches, and `false` otherwise.
+//
+func valueConformsToSemaType(value interpreter.Value, semaType sema.Type) (ok bool) {
+	switch typ := semaType.(type) {
+	case *sema.SimpleType:
+		ok = valueConformsToSimpleType(value, typ)
+	case *sema.IntType:
+		_, ok = value.(interpreter.IntValue)
+	case *sema.Int8Type:
+		_, ok = value.(interpreter.Int8Value)
+	case *sema.Int16Type:
+		_, ok = value.(interpreter.Int16Value)
+	case *sema.Int32Type:
+		_, ok = value.(interpreter.Int32Value)
+	case *sema.Int64Type:
+		_, ok = value.(interpreter.Int64Value)
+	case *sema.Int128Type:
+		_, ok = value.(interpreter.Int128Value)
+	case *sema.Int256Type:
+		_, ok = value.(interpreter.Int256Value)
+	case *sema.UIntType:
+		_, ok = value.(interpreter.UIntValue)
+	case *sema.UInt8Type:
+		_, ok = value.(interpreter.UInt8Value)
+	case *sema.UInt16Type:
+		_, ok = value.(interpreter.UInt16Value)
+	case *sema.UInt32Type:
+		_, ok = value.(interpreter.UInt32Value)
+	case *sema.UInt64Type:
+		_, ok = value.(interpreter.UInt64Value)
+	case *sema.UInt128Type:
+		_, ok = value.(interpreter.UInt128Value)
+	case *sema.UInt256Type:
+		_, ok = value.(interpreter.UInt256Value)
+	case *sema.Word8Type:
+		_, ok = value.(interpreter.Word8Value)
+	case *sema.Word16Type:
+		_, ok = value.(interpreter.Word16Value)
+	case *sema.Word32Type:
+		_, ok = value.(interpreter.Word32Value)
+	case *sema.Word64Type:
+		_, ok = value.(interpreter.Word64Value)
+	case *sema.Fix64Type:
+		_, ok = value.(interpreter.Fix64Value)
+	case *sema.UFix64Type:
+		_, ok = value.(interpreter.UFix64Value)
+	case *sema.OptionalType:
+		// Value must be `nil, or must be of type defined in the optional type.
+		if _, ok = value.(interpreter.NilValue); !ok {
+			ok = valueConformsToSemaType(value, typ.Type)
+		}
+	case *sema.AddressType:
+		_, ok = value.(interpreter.AddressValue)
+	case *sema.VariableSizedType:
+		ok = valueConformsToVariableSizedArrayType(value, typ)
+	case *sema.ConstantSizedType:
+		ok = valueConformsToConstantSizedArrayType(value, typ)
+	case *sema.DictionaryType:
+		ok = valueConformsToDictionaryType(value, typ)
+	case *sema.CompositeType:
+		ok = valueConformsToCompositeType(value, typ)
+	case *sema.CapabilityType:
+		_, ok = value.(*interpreter.CapabilityValue)
+
+	// Following types cannot be used as arguments to a script/transaction
+	case *sema.FunctionType:
+		_, ok = value.(interpreter.FunctionValue)
+	case *sema.TransactionType:
+		// false
+	}
+
+	return
+}
+
+func valueConformsToCompositeType(arg interpreter.Value, compositeType *sema.CompositeType) bool {
```

Comment:
Maybe already implicitly checked due to the dynamic check performed by the runtime after-wards, but given that this function should be independent on other usage, also check location and qualified identifier match